fl-server/src/serve_flists.rs

```diff
@@ -0,0 +1,207 @@
+use utoipa::ToSchema;
+
+use axum::{
+    body::Body,
+    http::{Request, StatusCode},
+    response::IntoResponse,
+};
+use axum_macros::debug_handler;
+use percent_encoding::percent_decode;
+
+#[debug_handler]
+pub async fn serve_flists(req: Request<Body>) -> impl IntoResponse {
+    let path = req.uri().path().to_string();
+
+    return match ServeDir::new("").oneshot(req).await {
+        Ok(res) => {
+            let status = res.status();
+            match status {
+                StatusCode::NOT_FOUND => {
+                    let path = path.trim_start_matches('/');
+                    let path = percent_decode(path.as_ref()).decode_utf8_lossy();
+
+                    let mut full_path = PathBuf::new();
+
+                    // validate
+                    for seg in path.split('/') {
+                        if seg.starts_with("..") || seg.contains('\\') {
+                            return Err(ErrorTemplate {
+                                err: ResponseError::BadRequest("invalid path".to_string()),
+                                cur_path: path.to_string(),
+                                message: "invalid path".to_owned(),
+                            });
+                        }
+                        full_path.push(seg);
+                    }
+
+                    let cur_path = std::path::Path::new(&full_path);
+
+                    match cur_path.is_dir() {
+                        true => {
+                            let rs = visit_dir_one_level(&full_path).await;
+                            match rs {
+                                Ok(files) => Ok(DirListTemplate {
+                                    lister: DirLister { files },
+                                    cur_path: path.to_string(),
+                                }
+                                .into_response()),
+                                Err(e) => Err(ErrorTemplate {
+                                    err: ResponseError::InternalError(e.to_string()),
+                                    cur_path: path.to_string(),
+                                    message: e.to_string(),
+                                }),
+                            }
+                        }
+                        false => Err(ErrorTemplate {
+                            err: ResponseError::FileNotFound("file not found".to_string()),
+                            cur_path: path.to_string(),
+                            message: "file not found".to_owned(),
+                        }),
+                    }
+                }
+                _ => Ok(res.map(axum::body::Body::new)),
+            }
+        }
+        Err(err) => Err(ErrorTemplate {
+            err: ResponseError::InternalError(format!("Unhandled error: {}", err)),
+            cur_path: path.to_string(),
+            message: format!("Unhandled error: {}", err),
+        }),
+    };
+}
+
+pub async fn visit_dir_one_level(path: &std::path::Path) -> io::Result<Vec<FileInfo>> {
```

Comment:
If you don't want to do `std::path::Path::new()` everywhere you could just rewrite this function signature as 

```suggestion
pub async fn visit_dir_one_level<P: AsRef<Path>>(path: P) -> io::Result<Vec<FileInfo>> {
   let path = path.as_ref();
```

this will make it possible to call this function as `visit_dir_one_level("some/path")`